multimixer-128/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+use core::fmt;
+
+use digest::crypto_common::{InvalidLength, Key, KeyInit, KeySizeUser};
+pub use digest::{self, Digest};
+
+use digest::typenum::Unsigned;
+use digest::{
+    consts::{U32, U64},
+    core_api::{
+        AlgorithmName, Block, BlockSizeUser, BufferKindUser, CoreWrapper, FixedOutputCore,
+        UpdateCore,
+    },
+    HashMarker, OutputSizeUser,
+};
+use rand_chacha::rand_core::{RngCore, SeedableRng};
+use rand_chacha::ChaCha8Rng;
+
+const BLOCKSIZE: usize = 32;
+
+#[derive(Clone)]
+pub struct MultimixerCore {
+    key_blocks: Vec<Block<Self>>,
+    block_sums: [u64; 8usize],
+    block_index: usize,
+    rng: Option<ChaCha8Rng>,
+}
+
+pub type Multimixer = CoreWrapper<MultimixerCore>;
+
+impl MultimixerCore {
+    fn compress(&mut self, message_block: &Block<Self>) {
+        let mut x: [u32; 4usize] = [0u32; 4];
+        //let mut h = [0u32; 4];
+        let mut y = [0u32; 4];
+        //let mut k = [0u32; 4];
+        let mut a = [0u32; 4];
+        let mut b = [0u32; 4];
+        let mut p = [0u32; 4];
+        let mut q = [0u32; 4];
+
+        let (h, k) = if let Some(ref mut rng) = self.rng.as_mut() {
+            let mut h = [0u32; 4];
+            let mut k = [0u32; 4];
+
+            for i in 0..4 {
+                h[i] = rng.next_u32();
+                k[i] = rng.next_u32();
+            }
+            (h, k)
+        } else {
+            let mut h = [0u32; 4];
+            let mut k = [0u32; 4];
+            for i in 0..4 {
+                h[i] = u32::from_ne_bytes([
+                    self.key_blocks[self.block_index][i * 4],
```

Comment:
What happens for messages bigger than key size? IIUC this code will panic.

Reply:
Yes, correct, actually the message and key should be equal in size. I first created the algorithm with a key, but then also implemented a PRNG with a seed. I really wanted to bench the code, and for this I would need a really big key. Ideally I would maybe create two different hash-classes, Multimixer-128 and Multimixer-128-PRNG. I think this would solve the other comments as well.